test/e2e/service_udn.go

```diff
@@ -0,0 +1,387 @@
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+
+	kapi "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2enode "k8s.io/kubernetes/test/e2e/framework/node"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2eservice "k8s.io/kubernetes/test/e2e/framework/service"
+)
+
+var _ = ginkgo.Describe("Network Segmentation: services", func() {
+	const (
+		serviceName       = "testservice"
+		servicePort       = 80
+		serviceTargetPort = 80
+		ovnWorkerNode     = "ovn-worker"
+		failDuration      = 3 * time.Second        // Duration for which failures must be consistent
+		failCheckInterval = 100 * time.Millisecond // Interval between checks
+	)
+
+	f := wrappedTestFramework("services-on-udn")
+
+	ginkgo.Context("on a user defined primary network", func() {
+		const (
+			userDefinedNetworkIPv4Subnet = "10.128.0.0/16"
+			userDefinedNetworkIPv6Subnet = "2014:100:200::0/60"
+			nadName                      = "tenant-red"
+			nodePort                     = 30125 // valid range is 30000-32767
+			nodePortDefaultService       = 30025
+		)
+
+		var (
+			cs                  clientset.Interface
+			nadClient           nadclient.K8sCniCncfIoV1Interface
+			clientNode          string
+			serverPodNodeName   string
+			defaultNetNamespace string
+		)
+
+		ginkgo.BeforeEach(func() {
+			cs = f.ClientSet
+
+			var err error
+			nadClient, err = nadclient.NewForConfig(f.ClientConfig())
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		})
+
+		cleanupFn := func() {
+			ginkgo.By("Removing the namespace so all resources get deleted")
+			err := cs.CoreV1().Namespaces().Delete(context.TODO(), f.Namespace.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "Failed to remove the namespace %s %v", f.Namespace.Name, err)
+			if defaultNetNamespace != "" {
+				err = cs.CoreV1().Namespaces().Delete(context.TODO(), defaultNetNamespace, metav1.DeleteOptions{})
+				framework.ExpectNoError(err, "Failed to remove the namespace %v", defaultNetNamespace, err)
+			}
+
+		}
+
+		ginkgo.AfterEach(func() {
+			cleanupFn()
+		})
+
+		ginkgo.DescribeTable(
+
+			// The test creates a nodeport service in a UDN backed by one pod and verifies that:
+			// - a client in the UDN on the same node as the backend pod can reach the service through the cluster IP
+			// - a client in the UDN can reach the service through the node port on all host IPs of the node where the client runs
+			// - a client in the UDN on a different node than the backend pod can reach the service through the cluster IP
+			// - a client in the default network cannot reach the service in the UDN through the cluster IP
+			// - a client in the default network cannot reach the service in the UDN on any host IPs
```

Comment:
we should add a test (maybe later if its not possible now) to make sure UDN pod cannot reach service in default network

Reply:
I have it already, it's at the end, but I forgot to update the comment you highlighted :(